### PR TITLE
feat: GPU shared texture offscreen rendering

### PR DESCRIPTION
**Pull request by reitowo**
#### Description of Change

I managed to add `ARGB + kGpuMemoryBuffer` support for FrameSinkVideoCapturer in [recent upstream changes](https://chromiumdash.appspot.com/commits?user=carolwolfking&platform=Windows). Thus, we can finally use zero-copy (actually one, there's a CopyRequest of frame texture) GPU shared texture OSR in chromium apps.

This will be the fastest OSR mode than the existing two, and it supports enabling hardware acceleration.

> However, this mode will not compose the popup widgets, but directly provide textures to user.

Details:
1. Add `webPreferences.offscreenUseSharedTexture` to enable this feature.
2. Add `texture` parameter to `paint` event of `webContent`.
3. Add structure definitions and docs.

Fix: #41972

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: feat: GPU shared texture offscreen rendering.

### CODE REVIEW

**Comment by reitowo** [comment]
Ready for review. Any suggestions? Here are some answers in advance:

1. Intentionally reused 'paint' event instead of creating a new one, because I don't see mandatory needs of creating a new event, and I prefer adding an extra parameter to 'paint'.

2. `offscreenUseSharedTexture` might be a little bit long, but I can't figure out a shorter name with same meaning.

3. Although we cannot do anything about the texture handle in pure node.js, we can write native node modules to import the texture, [example](https://github.com/reitowo/electron-spout)

4. Please confirm that `Emit('paint')` is a blocking call, as chromium will recycle the textures once the callback returns, so it has to be blocking.

5. The speed will be much faster than previous hardware accelerated one (~3ms vs <100us), and it has no more words about the software output device which has no GPU support and slow.

**Comment by reitowo** [comment]
Tested to be working on Windows, [example repo](https://github.com/reitowo/electron-spout-example)

[recording](https://github.com/electron/electron/assets/29846655/b904a262-2f21-42e7-8b47-793a32862a8f)

**Comment by Hate-Usernames** [comment]
> 3. Although we cannot do anything about the texture handle in pure node.js, we can write native node modules to import the texture, [example](https://github.com/reitowo/electron-spout)

Speaking as an interested outsider: how possible would it be to make it so that the texture could be used in the renderer process via webGPU with a workflow something like:

Main Process
```js
const appWin = new BrowserWindow({ width: 800, height: 800 })
appWin.loadFile('index.html')

const win = new BrowserWindow({ webPreferences: { offscreen: true, offscreenUseSharedTexture: true } })
win.webContents.on('paint', (event, dirty, image, texture) => {
  appWin.webContents.send('texture', texture)
})
win.loadURL('https://github.com')
```

Renderer Process
```js
ipcRenderer.on('texture', (_event, source) => {
  // Or similar function for chromium shared textures
  const externalTexture = gpuDevice.importExternalTexture({ source })
  // Use the texture in a webGPU pipeline and eventually
  // render the result to a canvas element
})
```

I'm biased as this is something I've been dreaming of for a long time, but if it is relatively simple to achieve, and performant, I think it could make this feature a lot more useful and accessible for the majority of the userbase.

**Comment by reitowo** [comment]
> renderer process via webGPU

Definitely possible. Actually I already did some research when I was walking through dawn's source. 

At least on Windows you can import a DXGI handle into dawn world (via webgpu native), then you can try convert it to a v8 webgpu texture to use it anywhere

**Comment by reitowo** [comment]
// PendingRemote instances may be freely moved to another thread/sequence, or
// even transferred to another process via a Mojo interface call (see
// pending_remote<T> syntax in mojom IDL).

I think ipcMain ipcRenderer doesn't support passing a pending_remote mojo? so I think the release callback it not able to pass to another process. The only reason will be supporting async on paint event to let you release the texture after some awaits.

**Comment by reitowo** [comment]
@Hate-Usernames Good news: I managed to let user able to release the texture when they want, which means you can pass the texture to wherever you want, but releasers need to be maintained in the callback process and cannot be ipc. I think it can already do what we want, nnaintaining that would be easy with few more ipc calls.

**Comment by reitowo** [comment]
I think the CI machines are dead for this PR? 🙂

File: docs/api/web-contents.md [hunk 1]
```diff
@@ -874,20 +874,47 @@ Returns:
...
 ```js
 const { BrowserWindow } = require('electron')
 
-const win = new BrowserWindow({ webPreferences: { offscreen: true } })
-win.webContents.on('paint', (event, dirty, image) => {
-  // updateBitmap(dirty, image.getBitmap())
+const win = new BrowserWindow({ webPreferences: { offscreen: true, offscreenUseSharedTexture: true } })
+win.webContents.on('paint', (event, dirty, image, texture) => {
+  if (texture) {
+    // importTextureHandle(dirty, texture.textureInfo)
```
Comment (nornagon) [inline docs/api/web-contents.md:888]: It would be great to have a little more info on how an app can actually use this. Perhaps in a separate guide doc?

File: docs/fiddles/features/offscreen-rendering/main.js [hunk 2]
```diff
@@ -9,13 +9,19 @@ function createWindow () {
...
+      offscreenUseSharedTexture: true // or false
     }
   })
 
   win.loadURL('https://github.com')
-  win.webContents.on('paint', (event, dirty, image) => {
-    fs.writeFileSync('ex.png', image.toPNG())
+  win.webContents.on('paint', (event, dirty, image, texture) => {
+    if (texture) {
+      // Import the shared texture handle to your own rendering world.
```
Comment (nornagon) [inline docs/fiddles/features/offscreen-rendering/main.js:20]: If you're changing the example to use shared textures by default, it should actually use them :)

File: patches/chromium/osr_shared_texture_remove_keyed_mutex_on_win_dxgi.patch [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+Subject: Remove DXGI GMB keyed-mutex
+
+This CL is only for reference for a vendor specific change. It is kept for rebasing.
+
+On CEF/electron, FrameSinkVideoCapturer is used in GPU accelerated OSR, and it can be faster
+if we remove the mutex on the shared resource for Windows.
+
+Maintain this patch in https://crrev.com/c/5465148
+
+Change-Id: Id90acd587f17acd228ae7cb5ef93005eb8388b80
```
Comment (nornagon) [inline patches/chromium/osr_shared_texture_remove_keyed_mutex_on_win_dxgi.patch:13]: Is this patch required for this feature? Why is this mutex there in the first place? Why can't this CL be merged upstream?

File: docs/api/web-contents.md [hunk 4]
```diff
@@ -874,20 +874,47 @@ Returns:
 * `event` Event
 * `dirtyRect` [Rectangle](structures/rectangle.md)
 * `image` [NativeImage](native-image.md) - The image data of the whole frame.
+* `texture` [OffscreenSharedTexture](structures/offscreen-shared-texture.md) - The GPU shared texture of the frame. Will be `null` if `webPreferences.offscreenUseSharedTexture` is `false`.
```
Comment (nornagon) [inline docs/api/web-contents.md:877]: Instead of extending the argument list, let's put stuff on the event. See https://github.com/electron/governance/blob/main/wg-api/best-practices.md#how-will-this-api-be-extended-in-the-future and https://github.com/electron/electron/pull/37085/files#diff-261075c0abf5bfd33be70313bcd3699da9b1222b6f27df0e8294fc63789f160bR1795-R1807 for an example. The C++ side of this is still a little rough but it beats long inscrutable argument lists.

**Review by nornagon — COMMENTED** [review]
This is interesting! I'm surprised that the code changes are so small. The API as written is a bit clunky, I've left some notes.

Also, I feel like it _ought_ to be possible to pass the texture handle across a mojo pipe to be released in another process, so if you're running into trouble with that I'm curious what the roadblock was!

Finally, OSR is quite a neglected feature of Electron, and a huge part of that is that the test coverage is quite poor. This PR adds a largish new API surface and will definitely need test coverage before it can be considered for merging.

**Comment by reitowo** [comment]
Hi! Thanks for the review.

> Also, I feel like it _ought_ to be possible to pass the texture handle across a mojo pipe to be released in another process, so if you're running into trouble with that I'm curious what the roadblock was!

According to the comments in chromium is possible to pass a `PendingRemote` mojo to another process, however forgive my unfamiliar with mojo, I thought it needs a explicit interface that declare such field to let it pass to another process with another receiver that accepts such parameter.

In this OSR senario, I passed most of shared texture info via to v8 by conversion, but I took a glance at electron's ipc and i think the serialization / deserialization process is not able to pass a `PendingRemote` object. So currently I wrapped the `PendingRemote` object in a structure pointer and save it in a v8 external, just passing it to local functions to let it release in the function callback. The `releaser` function is also not serializable so I really think it is not possible to pass purely at javascript side, without additional design in native side. 

In conclusion, it is possible to pass `PendingRemote` as a parameter of mojo interface in native side, but I find no clue about how to integrate such behaviour in electron's ipc to make it possible purely at javascript side. 

Hope you can help me out.

File: docs/api/structures/offscreen-shared-texture-releaser.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# OffscreenSharedTextureReleaser Object
+
+* `release` Function - Release the resources.
```
[the remark this answers is not in the document]
Comment (reitowo) [reply docs/api/structures/offscreen-shared-texture-releaser.md:3]: I find it difficult to declare in the docs: `Function\<Function>` ? `Function\<() => void\>` ? So I move it to a new Object.
Maybe I would better move the `OffscreenSharedTexture` from structures to classes as classes has `returns` section?

File: docs/api/structures/offscreen-shared-texture.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+To use in async or share this texture with another process, you need to call `takeReleaser` to take the ownership of the texture.
+  * `widgetType` string - The widget type of the texture, could be `popup` or `frame`.
+  * `pixelFormat` string - The pixel format of the texture, could be `rgba` or `bgra`.
+  * `sharedTextureHandle` string - _Windows_ _macOS_ The handle to the shared texture.
+  * `planes` Object[] - _Linux_ Each plane's info of the shared texture.
+    * `stride` number - The strides and offsets in bytes to be used when accessing the buffers via a memory mapping. One per plane per entry.
+    * `offset` number - The strides and offsets in bytes to be used when accessing the buffers via a memory mapping. One per plane per entry.
+    * `size` number - Size in bytes of the plane. This is necessary to map the buffers.
+    * `fd` number - File descriptor for the underlying memory object (usually dmabuf).
+  * `modifier` string - _Linux_ The modifier is retrieved from GBM library and passed to EGL driver.
```
[the remark this answers is not in the document]
Comment (reitowo) [reply docs/api/structures/offscreen-shared-texture.md:13]: According to chromium it will work on X11 and Wayland. However I only tested on Wayland but I think it is possible to work on any display server that supports gbm? These params are general on Linux.

File: docs/api/structures/offscreen-shared-texture.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+* `takeReleaser` Function\<[OffscreenSharedTextureReleaser](offscreen-shared-texture-releaser.md)> - Take the texture's ownership,
+prevent it being automatically released at callback end. The releaser cannot be passed to another process, users need to maintain
+releasers in current process, but it is safe to pass the `textureInfo` to another process.
```
[the remark this answers is not in the document]
Comment (reitowo) [reply docs/api/structures/offscreen-shared-texture.md:16]: Yeah, I agree with that. Its just a one line code, and user will not get more textures (pool max size) if they don't release. Maybe we provide some documenting.

File: docs/api/web-contents.md (hunk 1 above)
Comment (reitowo) [reply docs/api/web-contents.md:888]: It is currently very user dependent, personally I use a spout plugin [here](https://github.com/reitowo/electron-spout) to share the texture to external processes. So I think it maybe difficult to provide an pure electron example?
But good news is, it's possible to make WebGPU support importing such texture (with future works). However we probably need to find out how to pass `PendingRemote` to arbitrary processes as it needs to be imported at destination process. (Or not, you can always just pass the `textureInfo` and let the original process know to call the release).

File: docs/fiddles/features/offscreen-rendering/main.js (hunk 2 above)
Comment (reitowo) [reply docs/fiddles/features/offscreen-rendering/main.js:20]: Good suggestion! But like I described above, there's not a valid use purely inside electron, unless we managed to make it work with WebGPU.

File: patches/chromium/osr_shared_texture_remove_keyed_mutex_on_win_dxgi.patch (hunk 3 above)
Comment (reitowo) [reply patches/chromium/osr_shared_texture_remove_keyed_mutex_on_win_dxgi.patch:13]: It is not mandatory, but can provide an extra performance gain. Also as commented by OBS teams, texture mutex have performance spikes when acquiring them, and it may cause driver failure in extreme condition or old systems. Also according to the use case by OSR, the texture will not be written to again so a mutex is not requried. It is a tiny patch and I think it can be maintained easily, in general I think is worth to patch to eliminate performance spikes.
It is not possible to merge into upstream, as OSR is not the only use case in chromium, it was used as GMB (GpuMemoryBuffer) so a mutex is needed in some other code paths.

File: docs/api/web-contents.md (hunk 4 above)
Comment (reitowo) [reply docs/api/web-contents.md:877]: SGTM!

File: docs/fiddles/features/offscreen-rendering/main.js (hunk 2 above)
Comment (reitowo) [reply docs/fiddles/features/offscreen-rendering/main.js:20]: Of course, if you want to write it for testing I could write a testing node native module to copy the texture to a staging texture and read out the pixel data. 
However this is for Windows only, other systems may have similar approaches.

Comment (nornagon) [reply docs/fiddles/features/offscreen-rendering/main.js:20]: Testing would be the more important path, and I think it's OK to just test one platform at first.

For documentation, I think it would be better to leave this example to use the non-shared-texture path by default, perhaps with a comment about shared textures.

Comment (reitowo) [reply docs/fiddles/features/offscreen-rendering/main.js:20]: OK, I wi take a look about how to add a test, considering we need to compile (or just use precompiled binary?) it maybe complicated than other tests? It would be best if you could give some suggestions.

Comment (nornagon) [reply docs/fiddles/features/offscreen-rendering/main.js:20]: We have a couple of native modules in the repo already for testing this sort of thing. see [spec/fixtures/native-addon/echo](https://github.com/electron/electron/tree/main/spec/fixtures/native-addon/echo)

Comment (reitowo) [reply docs/fiddles/features/offscreen-rendering/main.js:20]: How do you suggest referencing a native-addon in spec/fixture to fiddles/feature?

**Comment by reitowo** [comment]
Hi @nornagon !

I've add tests and made change according to your review. Please do a second round review when you're vacant.

**Comment by reitowo** [comment]
Is there a special handler to serialize the texture object when passing through electron ipc? If so we might able to create a speicalized protocol to pass the ownership?

(Though I think managing purely at main process would also be an option as it is not complicated)

**Comment by admsev** [comment]
@gooroodev can you review this?

File: docs/fiddles/features/offscreen-rendering/main.js (hunk 2 above)
Comment (nornagon) [reply docs/fiddles/features/offscreen-rendering/main.js:20]: I think if there's no clear way to demonstrate the feature in a fiddle, then it shouldn't have a fiddle as documentation. It's okay for the feature to be documented in the docs and not have an accompanying fiddle.

File: patches/chromium/osr_shared_texture_remove_keyed_mutex_on_win_dxgi.patch (hunk 3 above)
Comment (nornagon) [reply patches/chromium/osr_shared_texture_remove_keyed_mutex_on_win_dxgi.patch:13]: Okay, that makes sense. Would you mind expanding the patch description with some of the info from this comment?

**Comment by reitowo** [comment]
@Hate-Usernames Good news, I've imported into WebGPU. However, I suspect if this will ever gonna merged into electron because of too hacky, lol.
image

**Comment by samuelmaddock** [comment]
What's the use case for supporting this in WebGPU? Is it something that could be covered by the [Element Capture API](https://developer.chrome.com/docs/web-platform/element-capture)? cc @Hate-Usernames

**Comment by Hate-Usernames** [comment]
> What's the use case for supporting this in WebGPU? Is it something that could be covered by the [Element Capture API](https://developer.chrome.com/docs/web-platform/element-capture)? cc @Hate-Usernames

As I understand it, the `getDisplayMedia` APIs discard the alpha channel, retaining this would be a key requirement for my use case (related to broadcast graphics).

**Comment by samuelmaddock** [comment]
> > What's the use case for supporting this in WebGPU? Is it something that could be covered by the [Element Capture API](https://developer.chrome.com/docs/web-platform/element-capture)? cc @Hate-Usernames
> 
> As I understand it, the `getDisplayMedia` APIs discard the alpha channel, retaining this would be a key requirement for my use case (related to broadcast graphics).

You may want to try describing your use case in the spec repo. They have an [issue for transparency support.](https://github.com/screen-share/element-capture/issues/16)

For Electron, it sounds like what you'd need would be a custom MediaStream constructed from the browser process. Discord added such a feature [in their fork.](https://github.com/discord/electron/pull/8) Something similar might be a more viable proposal for your use case.

**Comment by reitowo** [comment]
Thanks for the info.

There're also some senario when you want to import a texture completely from outside. However it doesn't work with WebGPU (at least SPEC version, not dawn or dawn native) currently.

File: docs/api/web-contents.md [hunk 8]
```diff
@@ -871,19 +871,33 @@ app.whenReady().then(() => {
 
 Returns:
 
-* `event` Event
+* `details` Event\<\>
+  * `texture` [OffscreenSharedTexture](structures/offscreen-shared-texture.md) - The GPU shared texture of the frame, when `webPreferences.offscreenUseSharedTexture` is `true`.
```
[the remark this answers is not in the document]
Comment (reitowo) [reply docs/api/web-contents.md:873]: Sure, I added experimental. However, GpuMemoryBufferHandle seems to be keeped or at least internal use last time I chat with them.

File: spec/api-browser-window-spec.ts [hunk 9]
```diff
@@ -6432,6 +6433,66 @@ describe('BrowserWindow module', () => {
     });
   });
 
+  describe('offscreen rendering image', () => {
```
[the remark this answers is not in the document]
Comment (reitowo) [reply spec/api-browser-window-spec.ts:6449]: Currently I'm busy for refactoring, may I refactor these next time if extra works need to be done?

File: docs/fiddles/features/offscreen-rendering/main.js (hunk 2 above)
Comment (reitowo) [reply docs/fiddles/features/offscreen-rendering/main.js:20]: Done.

File: shell/common/gin_converters/osr_converter.cc [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+                                    releaserHolder)
+                      .ToLocalChecked();
+
+  root.Set("release", releaser);
+
+  gin::Dictionary dict(isolate, v8::Object::New(isolate));
+  dict.Set("pixelFormat", OsrVideoPixelFormatToString(val.pixel_format));
+  dict.Set("widgetType", OsrWidgetTypeToString(val.widget_type));
+#if BUILDFLAG(IS_WIN) || BUILDFLAG(IS_MAC)
+  dict.Set("sharedTextureHandle", std::to_string(val.shared_texture_handle));
```
[the remark this answers is not in the document]
Comment (reitowo) [reply shell/common/gin_converters/osr_converter.cc:63]: Done

**Comment by reitowo** [comment]
@samuelmaddock @nornagon  Hi, guys. It's been a while cause I'm busy recently.

I've made changes according to your suggestions. Could you do another review? Hope we can merge this soon! :)

The bot closed this PR because I changed package.json, but it was mandatory for adding that fixture test.

File: docs/api/web-contents.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (reitowo) [reply docs/api/web-contents.md:873]: Nailed it.

File: docs/api/structures/offscreen-shared-texture.md [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# OffscreenSharedTexture Object
+
+* `textureInfo` Object - The shared texture info.
+  * `widgetType` string - The widget type of the texture, could be `popup` or `frame`.
+  * `pixelFormat` string - The pixel format of the texture, could be `rgba` or `bgra`.
+  * `sharedTextureHandle` string - _Windows_ _macOS_ The handle to the shared texture.
```
[the remark this answers is not in the document]
Comment (reitowo) [reply docs/api/structures/offscreen-shared-texture.md:6]: Done

**Comment by reitowo** [comment]
image

Also tested with `npm run test -- -g "offscreen\ rendering\ image"`

Hope you could see it soon.

**Comment by reitowo** [comment]
@codebytere Hi, Shelley.
Could you reopen this PR, and configure the bot to bypass for such situation that this PR adds an fixture test to `package.json`, which is essential, but triggered the bot policy?

**Comment by samuelmaddock** [comment]
@reitowo just want you to know I'm still watching this. Currently awaiting folks to take a closer look at #42855 so we can avoid PRs closing again. 

There's also an unfortunate limitation of GitHub we weren't aware of until now. After a GitHub PR closes, it [can't be reopened](https://github.com/isaacs/github/issues/361) if the underlying branch has been rebased. 🥲

Please hold off until the linked PR can be merged, then you'll need to create a new PR due to the limitation. We can pick up the review from there. Sorry about all this.

**Comment by reitowo** [comment]
Sure! 🥳

**Comment by samuelmaddock** [comment]
@reitowo the PR has been merged so you should be good to open up a new PR with these changes

**Comment by reitowo** [comment]
Done. See #42953
